File: backend/app/schemas/ai_chat.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator


class AIChatMessage(BaseModel):
    role: Literal["user", "assistant"]
    content: str = Field(min_length=1, max_length=4000)

    @field_validator("content")
    @classmethod
    def content_must_not_be_blank(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("content cannot be blank")
        return cleaned


class AIChatRequest(BaseModel):
    message: str = Field(min_length=1, max_length=4000)
    history: list[AIChatMessage] = Field(default_factory=list, max_length=20)
    web_search: bool = False
    conversation_id: str | None = Field(default=None, max_length=80)
    conversation_title: str | None = Field(default=None, max_length=4000)

    @field_validator("message")
    @classmethod
    def message_must_not_be_blank(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("message cannot be blank")
        return cleaned


class AIChatResponse(BaseModel):
    reply: str
    provider: str
    model: str
    usage: dict[str, Any]
    sources: list[dict[str, Any]] = Field(default_factory=list)
    latency_ms: int
    generation_record_id: int | None = None
    conversation_id: str
    conversation_title: str


class AIChatHistoryTurn(BaseModel):
    generation_record_id: int
    conversation_id: str
    conversation_title: str
    user_message: str
    assistant_message: str
    provider: str
    model: str
    web_search: bool = False
    latency_ms: int | None = None
    created_at: datetime


class AIChatConversationSummary(BaseModel):
    conversation_id: str
    title: str
    last_user_message: str
    last_assistant_message: str
    turn_count: int
    created_at: datetime
    updated_at: datetime

File: frontend-v2/src/lib/api/aiChat.ts
import { api } from "./client";

export type AIChatRole = "user" | "assistant";

export interface AIChatMessage {
  role: AIChatRole;
  content: string;
}

export interface AIChatResponse {
  reply: string;
  provider: string;
  model: string;
  usage: Record<string, unknown>;
  sources: Array<{ title?: string; url: string }>;
  latency_ms: number;
  generation_record_id: number | null;
  conversation_id: string;
  conversation_title: string;
}

export interface AIChatHistoryTurn {
  generation_record_id: number;
  conversation_id: string;
  conversation_title: string;
  user_message: string;
  assistant_message: string;
  provider: string;
  model: string;
  web_search: boolean;
  latency_ms: number | null;
  created_at: string;
}

export interface AIChatConversationSummary {
  conversation_id: string;
  title: string;
  last_user_message: string;
  last_assistant_message: string;
  turn_count: number;
  created_at: string;
  updated_at: string;
}

export async function sendAIChat(
  message: string,
  history: AIChatMessage[] = [],
  webSearch: boolean = false,
  conversationId?: string,
  conversationTitle?: string,
): Promise<AIChatResponse> {
  return api.post<AIChatResponse>("/api/ai-chat", {
    message,
    history,
    web_search: webSearch,
    conversation_id: conversationId,
    conversation_title: conversationTitle,
  });
}

export async function listAIChatHistory(limit: number = 20): Promise<AIChatHistoryTurn[]> {
  const params = new URLSearchParams();
  params.set("limit", String(limit));
  return api.get<AIChatHistoryTurn[]>(`/api/ai-chat/history?${params.toString()}`);
}

export async function listAIChatConversations(limit: number = 50): Promise<AIChatConversationSummary[]> {
  const params = new URLSearchParams();
  params.set("limit", String(limit));
  return api.get<AIChatConversationSummary[]>(`/api/ai-chat/conversations?${params.toString()}`);
}

export async function listAIChatConversationHistory(
  conversationId: string,
  limit: number = 50,
): Promise<AIChatHistoryTurn[]> {
  const params = new URLSearchParams();
  params.set("limit", String(limit));
  return api.get<AIChatHistoryTurn[]>(
    `/api/ai-chat/conversations/${encodeURIComponent(conversationId)}/history?${params.toString()}`,
  );
}
